include/envoy/network/listen_socket.h

```diff
@@ -34,5 +34,55 @@ class ListenSocket {
 typedef std::unique_ptr<ListenSocket> ListenSocketPtr;
 typedef std::shared_ptr<ListenSocket> ListenSocketSharedPtr;
 
+/**
+ * An abstract accepted socket.
+ */
+class AcceptedSocket {
+public:
+  virtual ~AcceptedSocket() {}
+
+  /**
+   * @return the address that the socket was received at, or an original destination address if
+   * applicable.
+   */
+  virtual Address::InstanceConstSharedPtr localAddress() const PURE;
+
+  /**
+   * Reset the destination address of the socket to a different address than the one
+   * the socket was accepted at.
+   */
+  virtual void resetLocalAddress(const Address::InstanceConstSharedPtr& local_address) PURE;
+
+  /**
+   * @return true if the local address has been reset.
+   */
+  virtual bool localAddressReset() const PURE;
+
+  /**
+   * @return the address that the socket was received from, or an original source address if
+   * applicable.
+   */
+  virtual Address::InstanceConstSharedPtr remoteAddress() const PURE;
+
+  /**
+   * Reset the source address of the socket to a different address than the one
+   * the socket was accepted at.
+   */
+  virtual void resetRemoteAddress(const Address::InstanceConstSharedPtr& remote_address) PURE;
+
+  /**
+   * @return fd the accepted socket's file descriptor.
+   */
+  virtual int fd() const PURE;
+
+  /**
+   * Transfer ownership of the file descriptor to the caller, so that the underlying socket will
+   * not be closed on delete.
+   */
+  virtual int takeFd() PURE;
```

Comment:
Per my other comment let's remove this from the exposed interface. I think we should be able to hide this as an implementation detail during listener processing (one option is to just store the `AcceptedSocket and actually pass it, vs. the `fd` to connections). If this is not easily possible can we add a TODO to think about cleaning this up? At the very least I don't want to pass this to filters, so we might need to have a split interface for internal/listener use and for filter use.

Reply:
OK, did this, but it was a lot of work and I had to bring close() back in to the exposed interface. Had to split up AcceptedSocket into multiple classes as it felt wrong to use that for ClientConnection. Let me know if you'd rather revert the last commit and do it as a separate PR after this one.